### ISSUE TITLE
bug: inference of strings with middleware defined after gets mangled

### ISSUE THREAD

**Issue by KATT** [trpc/trpc#5020]
### Provide environment information

latest

### Describe the bug

inference gets wrong, see below

### Link to reproduction

https://github.com/trpc/trpc/pull/5017

### To reproduce

```ts
import { inferRouterInputs, inferRouterOutputs, initTRPC } from '@trpc/server';
import { z } from 'zod';

const t = initTRPC.create();
const appRouter = t.router({
  str: t.procedure.input(z.string()).query(({ input }) => input),
  strWithMiddleware: t.procedure
    // ^?
    .input(z.string())
    .use((opts) => opts.next())
    .query(({ input }) => input),

  voidWithMiddleware: t.procedure
    .use((opts) => opts.next())
    .query(() => {
      // ..
    }),
});
type AppRouter = typeof appRouter;

describe('inferRouterInputs', () => {
  type AppRouterInputs = inferRouterInputs<AppRouter>;
  type AppRouterOutputs = inferRouterOutputs<AppRouter>;

  test('string', async () => {
    {
      type Input = AppRouterInputs['str'];
      type Output = AppRouterOutputs['str'];
      expectTypeOf<Input>().toBeString();
      expectTypeOf<Output>().toBeString();
    }
    {
      type Input = AppRouterInputs['strWithMiddleware'];
      //   ^? 

      type Output = AppRouterOutputs['strWithMiddleware'];
      expectTypeOf<Input>().toBeString();
      // ❌❌❌❌❌❌❌❌❌
      expectTypeOf<Output>().toBeString();
    }
  });
});

```

### Additional information

see #5017

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5020">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5020/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/5020/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
cc @jussisaurio would you like to have a look? 🙏 ❤️

**Comment by jussisaurio** [comment]
Pushed a tentative fix for the `Overwrite` utility function, lmk what you think

**Comment by jussisaurio** [comment]
Pushed to your draft PR, that is :)

**Comment by juliusmarminge** [comment]
Closed by #5017
